Blogpost on choosing good chunk sizes in Dask (#119)
dask/dask-blog: _posts/2021-11-02-choosing-dask-chunk-sizes.md

---
layout: post
title: Choosing good chunk sizes in Dask
author: Genevieve Buckley
tags: [performance]
theme: twitter
---
{% include JB/setup %}

## Summary

Confused about choosing [a good chunk size](https://docs.dask.org/en/latest/array-best-practices.html#select-a-good-chunk-size) for Dask arrays?

Array chunks can't be too big (we'll run out of memory), or too small (the overhead introduced by Dask becomes overwhelming). So how can we get it right?

It's a two step process:

> 1. First, start by choosing a chunk size similar to data you know can be processed entirely within memory (i.e. without dask)

Reviewer: ```suggestion
1. First, start by choosing a chunk size similar to data you know can be processed entirely within memory (i.e. without Dask)
```

> 2. Then, watch the dask dashboard task stream and worker memory plots, and adjust if needed.

Reviewer: ```suggestion
2. Then, watch the Dask dashboard task stream and worker memory plots, and adjust if needed.
```

## Contents

- [ What are Dask array chunks?](#what-are-dask-array-chunks)
- [Choosing an initial chunk size](#choosing-an-initial-chunk-size)
- [Too small is a problem](#too-small-is-a-problemg)
- [Too big is also a problem](#too-big-is-also-a-problem)
- [Using the Dask dashboard](#using-the-Dask-dashboard)
- [Unmanaged memory](#unmanaged-memory)
- [Thanks for reading](#thanks-for-reading)

## What are Dask array chunks?

Dask arrays are big structures, made out of many small chunks.
Typically, each small chunk is an individual [numpy array](https://numpy.org/), and they are arranged together to make a much larger Dask array.

<img src="https://raw.githubusercontent.com/dask/dask/ac01ddc9074365e40d888f80f5bcd955ba01e872/docs/source/images/dask-array-black-text.svg" alt="Diagram: Dask array chunks" width="400" height="300" />

You can find more information about Dask array chunks on this page of the documentation: https://docs.dask.org/en/latest/array-chunks.html

> ### How do I nkow what chunks my array has?

Reviewer: ```suggestion
### How do I know what chunks my array has?
```

If you have a Dask array, you can use the `chunksize` or `chunks` attribues to see information about the chunks. You can also visualize this with the Dask array HTML representation.

![Visualizating Dask array chunks with the HTML repr](/images/choosing-good-chunk-sizes/examine-dask-array-chunks.png)

`arr.chunksize` shows the largest chunk size. For arrays where you expect roughly uniform chunk sizes, this is a good way to summarize chunk size information.

`arr.chunks` shows fully explicit sizes of all chunks along all dimensions within the Dask array (see [item 3 here](https://docs.dask.org/en/stable/array-chunks.html#specifying-chunk-shapes)). This is more verbose, and is a good choice with arrays that have irregular chunks.

## Choosing an initial chunk size

Some very rough rules of thumb:

> - When developing a new data processing pipeline, many people create a prototype using a small subset of the data, which may not involve Dask at all. This is helpful for quickly iterating on ideas, and also has the secondary benefit of giving you a clear idea of what size data can be processed easily for this workflow. You can use this knowledge to choose similar sized chunks in Dask.

Reviewer: ```suggestion
- If you already created a prototype, which may not involve Dask at all, using a small subset of the data you intend to process, you'll have a clear idea of what size of data can be processed easily for this workflow. You can use this knowledge to choose similar sized chunks in Dask.
```

Author: Much clearer! Thank you very much

> - Some people have observed that chunk sizes below 1MB are almost always bad. Chunk size between 50MB and 1GB are generally good (and chunk sizes over 2GB may be too big - but this upper limit varies a huge amount depending on the computational resources you have available, so adjust for your circumstances)

Reviewer: The link given at the beginning of the article even say "it is rare to see chunk sizes below 100 MB". Not sure how to bring this here, or if we should...

Author: So perhaps this should say instead "chunk sizes between 100MB and 1GB are generally good"

> - Some people have observed that chunk sizes below 1MB are almost always bad. Chunk size between 50MB and 1GB are generally good (and chunk sizes over 2GB may be too big - but this upper limit varies a huge amount depending on the computational resources you have available, so adjust for your circumstances)

Reviewer: ```suggestion
- Some people have observed that chunk sizes below 1MB are almost always bad. Chunk size between 50MB and 1GB are generally good, going over 1 or 2GB means you have a really big dataset and/or a lot of memory available per core,
```

> - Upper bound: Avoid too very large task graphs. More than 10,000 or 100,000 chunks will perform poorly for this reason.

Reviewer: ```suggestion
- Upper bound: Avoid too large task graphs. More than 10,000 or 100,000 chunks may start to perform poorly.
```

> - Lower bound: To get the advantage of parallization, you need the number of chunks to at least equal the number of workers available (or better, the number of workers times 2). Otherwise, if the number of chunks is less than the number of workers, some workers will stay idle.

Reviewer: ```suggestion
- Lower bound: To get the advantage of parallelization, you need the number of chunks to at least equal the number of worker cores available (or better, the number of worker cores times 2). Otherwise, some workers will stay idle.
```

Some examples of data storage structures on disk include:
- A HDF5 or [Zarr array](https://zarr.readthedocs.io/en/stable/api/core.html). The size and shape of chunks/blocks stored on disk should align well with the Dask array chunks you select.
- A folder full of tiff files. You might decide that each tiff file should become a single chunk in the Dask array (or that multiple tiff files should be grouped into a single chunk).

### Rechunking arrays

If you want to change the chunking of a Dask array, you can do that with the [rechunk](https://docs.dask.org/en/latest/generated/dask.array.rechunk.html) method.

```python
rechunked_array = original_array.rechunk(new_shape)
```

**Warning:** Rechunking Dask arrays comes at a cost.
- The Dask graph must be rearranged to accomodate the new chunk structure. This happens immediately, and will block any other interaction with python until Dask has rearranged the task graph.
- This also inserts new tasks into the Dask graph. At compute time, there are now more tasks to execute.

For these reasons, it is best to avoid rechunking. However, sometimes the data is stored on disk is not well aligned and rechunking may be necessary.

> For an example of this, here is Draga Doncila Pop [talking about chunk alignment](https://youtu.be/10Ws59NGDaE?t=833) with satellite image data.

Reviewer: Should we introduce [rechunker](https://github.com/pangeo-data/rechunker), which has been built to overcome the complexity of rechunking large dask arrays in memory?

Author: Oh yes, I think that would be good to include.

Author: Added:
> The [rechunker](https://rechunker.readthedocs.io/en/latest/) library can be useful in these situations:
>
> > Rechunker takes an input array (or group of arrays) stored in a persistent storage device (such as a filesystem or a cloud storage bucket) and writes out an array (or group of arrays) with the same data, but different chunking scheme, to a new location. Rechunker is designed to be used within a parallel execution framework such as Dask.

> ## Too small is a problem

Reviewer: I'll would put this part and the one after just below the "How do I nkow what chunks my array has?" part, or even just before.

Author: Section order rearranged. I think that is better, thank you

If array chunks are too small, it's inefficient. Why is this?

Using Dask introduces some amount of overhead for each task in your computation.
This overhead is the reason the Dask best practices advise you to [avoid too-large graphs](https://docs.dask.org/en/latest/best-practices.html#avoid-very-large-graphs).
This is because if the amount of actual work done by each task is very tiny, then the percentage of overhead time vs useful work time is not good.

> Typically, the Dask scheduler takes 1 millisecond to coordinate a single task. That means we want the computation time for each task to be comparitively larger, eg: seconds instead of milliseconds.

Reviewer: This is already said above. But I would remove the other one and keep this one.

Author: I agree that this is important to keep here (and now with the rearranging of sections, this part comes first in the article). 

But I don't want to remove it from the rules of thumb list for two reasons:
1 It is a very clear and specific guideline so I think it belongs, and
2. Most people will assume the "rough rules of thumb" list is a good summary of the takeaways from this article, so if we omit it from the list people will likely forget or overlook this point.
I think a small amount of repetition is ok in this case.

Instead, we can do this in a smarter way. The foreman (Dask scheduler) can tell the workers to go and bring one full wheelbarrow load of bricks back each time. Now workers are spending much less time moving between the wall and the pile of bricks, and the wall will be finished much quicker.

## Too big is also a problem

If the Dask array chunks are too big, this is also bad. Why?
Chunks that are too large are bad because then you are likely to run out of working memory.
You may see out of memory errors happening, or you might see performance decrease substantially as data spills to disk.

When too much data is loaded in memory on too few workers, Dask will try to spill data to disk instead of crashing.
Spilling data to disk makes things run very slowly, because all the extra read/write operations to disk. Things don't just get a little bit slower, they get a LOT slower, so it's smart to watch out for this.

> To avoid data being spilled to disk, watch the **worker memory plot** on the dask dashboard.

Reviewer: ```suggestion
To avoid data being spilled to disk, watch the **worker memory plot** on the Dask dashboard.
```

- [Intro to the Dask dashboard (18 minute video)](https://www.youtube.com/watch?v=N_GqzcuGLCY)
- [Dask Jupyterlab extension (6 minute video)](https://www.youtube.com/watch?v=EX_voquHdk0)
- [Dask dashboard documentation](http://distributed.dask.org/en/latest/diagnosing-performance.html)

We recommend always having the dashboard up when you're working with Dask.
It's a fantastic way to get a sense of what's working well, or poorly, so you can make adjustments.

Bad signs to watch out for include:
- Lots of white space in the task stream plot is a bad sign. White space means nothing is happening. Chunks may be too small.
- Lots and lots of red in the task stream plot is a bad sign. Red means worker communication. Dask workers need some communication, but if they are doing almost nothing except communication then there is not much productive work going on.
- On the worker memory plot, watch out for orange bars which are a sign you are getting close to the memory limit. Chunks may be too big.

> - On the worker memory plot, watch out for grey bars which mean data is being spilled to disk. Chunks may be too big.

Reviewer: Maybe sum-up the last two points as already explained in previous chapter?

Author: Probably better to keep this and remove the other one. We can link to this from that other section.

Author: Upon further review, my opinion is the same as above. Now that we have rearranged the order of sections, I think the small amount of repetition here is helpful to reinforce the key points.
